### DISCUSSION TITLE
Submaps pipeline and coordinate systems documentation

### DISCUSSION THREAD

**Discussion by twdragon** [IGMR-RWTH/RaNDT-SLAM#5] [General]
The idea of this kind of SLAM solution seems very interesting, especially in terms of the trajectory-based segmentation of the scan matcher workspace, which idea has looked really brilliant since [Google Cartographer](https://github.com/cartographer-project/cartographer) emerged. From the developer's point of view, there is also a very interesting and challenging problem with the usage of predefined segmented geometry as the basis of the map. Due to this, it could be very useful to have (not only in the paper but also in the public wiki) the following information:
- A detailed (possibly graphic) explanation of all the coordinate systems connected to the submap, and the reference transformations between them.
- Structure of the pose graph which is not documented at all.
- Explanation about how the decision is made about changing from one submap to another during the localization.
- Reference structure information for the pose graph to analyze its structure and artificialize it during loading of the predefined vector graphics into the system.

**Comment by nilsmandischer** [comment]
Hi @twdragon, I moved the issue to discussion as it has more of a Q&A character. You will already find some answers in the paper ([arXiv](https://www.arxiv.org/abs/2408.11576)). @maxhilger has more insight into the pose graph, he can most likely answer your question.

**Comment by maxhilger** [comment]
Hi @twdragon ,
sorry for my late reply. Now that IROS is over, I have more time.
To briefly answer your questions:
* The origin of each submap is stored in the "origin_in_global_frame_" variable. In each submap, the trajectory is saved in the "trajectory_" variable, referenced to the submap origin
* The pose graph is made up of all keyframes. This includes the first frame of each submap, and every "insertion_step"th scan. A more elegant way to handle this would probably be to replace this keyframe logic with a metric threshold, e.g., every 1 m or 5 deg.
* First of all, we currently do not have localization mode. As for the keyframes, a submap is initialized after processing a fixed number "submap_size_poses". If you want to implement a localization mode, I'd use a combination of ScanContext and a similarity measure given the last pose estimate to look for candidate submaps
* We currently cannot serialize or deserialize our pose graph, but that is aimed at being added within the near future. Note that the OGM map generated in this package is not used for pose estimation but is purely added to enable simple downstream tasks.

I hope that this answers your questions, feel free to circle back if I should elaborate more.
